### PR TITLE
[3.9] gh-92112: Fix crash triggered by an evil custom `mro()` (GH-92113)

### PR DESCRIPTION
**Pull request by JelleZijlstra**
(cherry picked from commit 85354ed78c0edb6d81a2bd53cabc85e547b8b26e)

Co-authored-by: Alexey Izbyshev <izbyshev@ispras.ru>

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -317,25 +317,29 @@ type_mro_modified(PyTypeObject *type, PyObject *bases) {
...
 
     if (!_PyType_HasFeature(type, Py_TPFLAGS_HAVE_VERSION_TAG))
         return;
 
     if (custom) {
+        PyObject *mro_meth, *type_mro_meth;
         mro_meth = lookup_maybe_method(
-            (PyObject *)type, &PyId_mro, &unbound);
-        if (mro_meth == NULL)
+            (PyObject *)type, &_Py_ID(mro), &unbound);
```
Comment (JelleZijlstra) [inline Objects/typeobject.c]:
```suggestion
            (PyObject *)type, &PyId_mro, &unbound);
```

File: Objects/typeobject.c [hunk 2]
```diff
@@ -317,25 +317,29 @@ type_mro_modified(PyTypeObject *type, PyObject *bases) {
...
+        if (mro_meth == NULL) {
             goto clear;
+        }
         type_mro_meth = lookup_maybe_method(
-            (PyObject *)&PyType_Type, &PyId_mro, &unbound);
-        if (type_mro_meth == NULL)
+            (PyObject *)&PyType_Type, &_Py_ID(mro), &unbound);
```
Comment (JelleZijlstra) [inline Objects/typeobject.c]:
```suggestion
            (PyObject *)&PyType_Type, &PyId_mro, &unbound);
```

**Comment by ambv** [comment]
Doesn't look like the fix actually succeeded?

**Comment by JelleZijlstra** [comment]
Yeah I think I messed up the backport and I procrastinated fixing it. I can look into it tonight (California time), though maybe that's too late for the next 3.9 release.

**Comment by ambv** [comment]
The release I'll be cutting tonight is the last 3.9 bugfix release.

**Comment by ambv** [comment]
Thanks, @JelleZijlstra!

**Comment by JelleZijlstra** [comment]
Thanks for checking in! Fortunately it was a simple fix.

**Pull request merged**